File: 155-min-stack/min-stack.py
class MinStack:

    def __init__(self):
        self.stack = []
        self.minStack = []
        

    def push(self, val: int) -> None:
        self.stack.append(val)
        # if minStack is empty *or* val is <= current min, push it
        if not self.minStack or val <= self.minStack[-1]:
            self.minStack.append(val)
        

    def pop(self) -> None:
        if self.isEmpty():
            return None
        value = self.stack.pop()
        if self.minStack and self.minStack[-1] == value:
            self.minStack.pop()
        return value
        

    def top(self) -> int:
        if self.isEmpty():
            return None
        value = self.stack[-1]
        return value
        

    def getMin(self) -> int:
        if self.isEmpty():
            return None
        value = self.minStack[-1]
        return value
    
    def isEmpty(self) -> bool:
        if len(self.stack) == 0:
            return True
        return False
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 155-min-stack/min-stack.java


import javax.xml.crypto.dsig.keyinfo.RetrievalMethod;

class MinStack {

    Stack<Integer> minStack;
    Stack<Integer> stack;
    
    public MinStack() {

        minStack = new Stack<>();
        stack = new Stack<>();

    }
    
    public void push(int val) {

        stack.add(val);
        if (minStack.size() == 0 || minStack.peek() >= val) {
            minStack.add(val);
        }
    }
    
    public void pop() {

        int popVal = stack.pop();
        if (minStack.size() > 0 && minStack.peek().equals(popVal)) {
            minStack.pop();
        }
    }
    
    public int top() {

        return stack.peek();
        
    }
    
    public int getMin() {

        return minStack.peek();
        // return stack.peek();
        
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */